### PR TITLE
router: support prefix wildcards in virtual hosts domains

### LINKED ISSUE #1269: Teach domains to allow trailing wildcards / unknown ports in Host
The "domains" configuration within a virtual_hosts of a route config allows wildcards at the start of the domain currently, but not the tail.

This means If I do something (like for local development) where I run envoy on port 8080 locally with a name based virtual host / domains set (ideally the same one i'm going to be deploying), I have to match the port exactly inside the domains, otherwise the request doesn't match and I get a 404 from envoy, which has no immediate debugging feedback as to why it's happening

Open to writing up a patch to help implement this, but not sure what's the approach that would be preferred:
1. Match domains excluding the port
2. Allow "trailing" wildcards, so things like `example.com:80*`, `example.com:*` would properly function

Some other option I haven't thought of here.

### PR DESCRIPTION
**Pull request by bartebor**
Description: Adds prefix wildcard support (`foo.*`) in virtual host domains
Risk Level: Low (does not change current behavior)
Testing: Unit tests
Docs Changes: updated `domains` field documentation in proto file
Release Notes: updated
Fixes #1269

### CODE REVIEW

File: docs/root/intro/version_history.rst [hunk 1]
```diff
@@ -53,6 +53,7 @@ Version history
 * router: added reset reason to response body when upstream reset happens. After this change, the response body will be of the form `upstream connect error or disconnect/reset before headers. reset reason:`
 * router: added :ref:`rq_reset_after_downstream_response_started <config_http_filters_router_stats>` counter stat to router stats.
 * router: added per-route configuration of :ref:`internal redirects <envoy_api_field_route.RouteAction.internal_redirect_action>`.
+* router: added support for prefix wildcards in :ref:`virtual hosts domains<envoy_api_field_route.VirtualHost.domains>`
```
[the remark this answers is not in the document]
Comment (bartebor) [reply docs/root/intro/version_history.rst]: Thanks @venilnoronha, I fixed this. I made the same mistake in commit message which would need to be taken care of when squashing (I did not rewrite history).

**Comment by htuch** [comment]
@junr03 doesn't have to be a trie in this PR due to the fact we only do a list already (Trevor added a trie a while back but then reverted). What's important to me is that it is plausible to implement the semantics being added in a trie (and inverted trie). This is because we want to add scalability in the future, and so need to ensure we're not adding anything to VH matching that would prohibit this (e.g. if we added a regex, I would be opposed to that).

File: api/envoy/api/v2/route/route.proto [hunk 2]
```diff
@@ -39,17 +39,21 @@ message VirtualHost {
...
 
   // A list of domains (host/authority header) that will be matched to this
-  // virtual host. Wildcard hosts are supported in the form of ``*.foo.com`` or
-  // ``*-bar.foo.com``.
+  // virtual host. Wildcard hosts are supported in the suffix or prefix form.
+  //
+  // Domain search order:
+  //  1. Exact domain names: ``www.foo.com``.
+  //  2. Suffix domain wildcards: ``*.foo.com`` or ``*-bar.foo.com``.
+  //  3. Prefix domain wildcards: ``foo.*`` or ``foo-*``.
```
Comment (htuch) [inline api/envoy/api/v2/route/route.proto]: How does this work with SNI? Presumably if we have wildcard certs, they won't work as prefix domains, so you will need to have exact certs or suffix domain certs with SAN. That's a completely different part of Envoy, but I'm wondering operationally whether prefix domain matching will only really work for non-TLS downstream.

Comment (bartebor) [reply api/envoy/api/v2/route/route.proto]: @htuch AFAIK prefix wildcard certs will/should not work with SNI. Although it is technically possible to create a certificate with a name, say, `mail.*`, I doubt any trusted CA would issue such certificate due to security reasons. I could imagine such scenario with own CA only. I have made such experiment using nginx/curl to no success.

Having said that I think SNI and HTTP domain matching are quite loosely coupled. One can have listener(s) configured with or without suffix wildcards, with or without TLS and at the same time match vhosts with prefix wildcards. It depends on what suits best given scenario.

**Comment by bartebor** [comment]
@junr03 I chose the same order nginx uses. I believe the idea behind this order is matching with steps of decreasing selectivity. Prefix wildcards can match any domain, so they are matched last. Nginx additionally supports regexps as a last step, but we probably do not want that for the reasons @htuch mentioned in https://github.com/envoyproxy/envoy/pull/6303#issuecomment-474553923.

File: api/envoy/api/v2/route/route.proto (hunk 2 above)
Comment (htuch) [reply api/envoy/api/v2/route/route.proto]: Sure, I agree on the loose coupling, just wondering out loud about how an operator would make use of these two features in conjunction. Nothing actionable, so I will resolve this thread.

File: source/common/router/config_impl.cc [hunk 3]
```diff
@@ -977,6 +979,9 @@ RouteMatcher::RouteMatcher(const envoy::api::v2::RouteConfiguration& route_confi
         default_virtual_host_ = virtual_host;
       } else if (domain.size() > 0 && '*' == domain[0]) {
         wildcard_virtual_host_suffixes_[domain.size() - 1].emplace(domain.substr(1), virtual_host);
+      } else if (domain.size() > 0 && '*' == domain[domain.size() - 1]) {
+        wildcard_virtual_host_prefixes_[domain.size() - 1].emplace(
+            domain.substr(0, domain.size() - 1), virtual_host);
```
Comment (junr03) [inline source/common/router/config_impl.cc]: Same with the suffixes above. But I just noticed that we don't detect duplicates (at least afaict here) for suffix or prefix domains, like we do for exact matches below. Is that intentional?

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -1012,7 +1017,8 @@ RouteConstSharedPtr VirtualHostImpl::getRouteFromEntries(const Http::HeaderMap&
 
 const VirtualHostImpl* RouteMatcher::findVirtualHost(const Http::HeaderMap& headers) const {
   // Fast path the case where we only have a default virtual host.
-  if (virtual_hosts_.empty() && wildcard_virtual_host_suffixes_.empty() && default_virtual_host_) {
+  if (virtual_hosts_.empty() && wildcard_virtual_host_suffixes_.empty() &&
+      wildcard_virtual_host_prefixes_.empty() && default_virtual_host_) {
```
Comment (junr03) [inline source/common/router/config_impl.cc]: I wonder if we even need the `default_virtual_host_` clause here. We end up returning `default_virtual_host_.get()` at the end anyway.

File: source/common/router/config_impl.cc (hunk 3 above)
Comment (bartebor) [reply source/common/router/config_impl.cc]: I don't personally think it was intentional. I'll add duplicate checking to make things clear.

File: source/common/router/config_impl.cc (hunk 4 above)
Comment (bartebor) [reply source/common/router/config_impl.cc]: You're right, clause removed.

**Comment by junr03** [comment]
@htuch can you do a final pass?

@bartebor do you mind fixing the conflict?

File: test/common/router/config_impl_test.cc [hunk 5]
```diff
@@ -315,8 +318,16 @@ TEST_F(RouteMatcherTest, TestRoutes) {
             config.route(genHeaders("bar.baz.com", "/", "GET"), 0)->routeEntry()->clusterName());
   EXPECT_EQ("instant-server",
             config.route(genHeaders(".foo.com", "/", "GET"), 0)->routeEntry()->clusterName());
-  EXPECT_EQ("instant-server",
+  EXPECT_EQ("wildcard",
```
Comment (htuch) [inline test/common/router/config_impl_test.cc]: Are these tests strong enough to show that the match order follows the specified one (suffix, then prefix wildcards)?

Comment (bartebor) [reply test/common/router/config_impl_test.cc]: @htuch I refactored my tests to handle all cases explicitly.

Something bad is going with the tests from the point I merged with master. I don't think failures result from my code changes, but with no access to test artifacts I can't be sure.

**Pull request merged**